File: problems-extend-adapt/1401-1500/1485_duplicate-binary-tree-with-random-links/solution_clone_map.js
/**
 * @param {RandomTreeNode} root
 * @return {RandomTreeNode}
 */
var duplicateRandomLinkedTree = function (root) {
    const clones = new Map();
    const clone = (node) => {
        if (node === null) return null;
        if (clones.has(node)) return clones.get(node);
        const copy = new RandomTreeNode(node.val);
        clones.set(node, copy);
        copy.left = clone(node.left);
        copy.right = clone(node.right);
        copy.random = clone(node.random);
        return copy;
    };
    return clone(root);
};

File: problems/1401-1500/1485_duplicate-binary-tree-with-random-links/solution_clone_map.java
import java.util.HashMap;
import java.util.Map;

class Solution {

    private Map<RandomTreeNode, RandomTreeNode> copies = new HashMap<>();

    public RandomTreeNode duplicateRandomLinkedTree(RandomTreeNode root) {
        if (root == null) {
            return null;
        }
        RandomTreeNode existing = copies.get(root);
        if (existing != null) {
            return existing;
        }
        RandomTreeNode copy = new RandomTreeNode(root.val);
        copies.put(root, copy);
        copy.left = duplicateRandomLinkedTree(root.left);
        copy.right = duplicateRandomLinkedTree(root.right);
        copy.random = duplicateRandomLinkedTree(root.random);
        return copy;
    }
}
